### ISSUE TITLE
Testing Alt - Best Practices

### ISSUE THREAD

**Issue by srph** [goatslacker/alt#153]
\* Didn't post this in Slack or Gitter because I felt this would be a little long.

I got a few questions I thought I should put here, so in case other people gets the same questions to themselves. I'm fairly new to testing, and especially to Flux. For any obvious stuff, please just point it out, thanks :sparkles:.

I've also seen Flux's note on testing, but I'm kind of confused due to the small differences. Not sure if I should do X or Y.

- I haven't tried mocking out components. If my wrapper component uses `AltContainer`, how do I test it? Do I mock the said component or anything? Here's an [example](https://github.com/sourcescript/alt-notify/blob/master/src/Drawer.jsx#L49). I've tested it, but I only stubbed out the store's `getState` method as shown [here](https://github.com/sourcescript/alt-notify/blob/master/src/__tests__/Drawer-spec.js#L9).

- Related to the item above. What if the wrapper component exposes a method (which calls an Action to its children through props), how do I properly test it as shown [here](https://github.com/sourcescript/alt-notify/blob/master/src/Drawer.jsx#L59)? Which do I *stub* / *mock* besides the `store`'s `getState`? Maybe I should use **sinon**'s [`onFirstCall`](http://sinonjs.org/docs/#stub-onCall) or [`onSecondCall`](http://sinonjs.org/docs/#stub-onCall)?

- If my Actions do nothing much expect dispatch data as shown [here](https://github.com/sourcescript/alt-notify/blob/master/src/NotifyActions.js), do I need to do much testing? Is it alright for Actions to transform data instead of the Store like in this [example](https://github.com/sourcescript/alt-notify/blob/master/src/NotifyStore.js#L17)?

If this is unrelated, feel free to close. Thanks!

**Comment by jdlehman** [comment]
TLDR; 
- [alt testing utils](https://github.com/goatslacker/alt/blob/master/utils/AltTestingUtils.js)
- [alt testing docs](https://github.com/goatslacker/alt/tree/master/docs/testing)

You have posed a few tricky questions regarding testing that don't necessarily have a right/wrong answer.  The amount people like to mock/stub objects varies.  Some people do not like to use any mocking.  I typically try to limit mocking/stubbing except where I really need it/ if it makes testing something much easier.

- Stubbing the store's `getState` is a great way to test components that rely on stores because it allows you to isolate such that you are only testing the component.  I don't think you need to stub the `AltContainer` in your example, it is a part of the overall component you are testing.  If you refactored the JSX the container is wrapping into its own component and file, then you could test it by itself, but that is a preference.
- To test component behavior you can use React's [TestUtils](http://facebook.github.io/react/docs/test-utils.html) to simulate DOM events.  If the handler changes the state of the component, you can then have an assertion to check that the state has changed correctly.  If it does other things such as call an Action (like in your example), sinon is a great tool to stub that Action and assert that it was `calledOnce` or `calledWithArgs`.
- If your Actions do not do anything but pass data on to the dispatcher, there is no need for a test as you aren't really testing anything (except maybe typos).  If you are more comfortable having tests for everything go for it, there is nothing wrong with it.  You could also clean up some of the boilerplate using [`generateActions`](https://github.com/goatslacker/alt/blob/master/docs/createActions.md#generateactions).

Hopefully that is helpful!

**Comment by srph** [comment]
I forgot about my question. Anyway, thanks, @jdlehman :+1: :sparkles:.

**Comment by goatslacker** [comment]
We really need to expand the docs on testing.

**Comment by dphaener** [comment]
@goatslacker I'm working with @zackify to hopefully create some middleware for legit-tests.

**Comment by goatslacker** [comment]
Sweeeet
